src/datahike/db.cljc

```diff
@@ -1560,20 +1557,23 @@
       true    (update :hash + (hash datom))
       schema? (-> (update-schema datom) update-rschema))))
 
-(defn- transact-report-upsert [report datom]
-  (let [db      (:db-after report)
-        a       (:a datom)
-        report' (-> report
-                    (update-in [:db-after] with-datom-upsert datom)
-                    (update-in [:tx-data] conj datom))]
-    (if (tuple-source? db a)
-      (let [e      (:e datom)
-            v      (if (datom-added datom) (:v datom) nil)
-            queue  (or (-> report' ::queued-tuples (get e)) {})
-            tuples (get (-attrs-by db :db/attrTuples) a)
-            queue' (queue-tuples queue tuples db e v)]
-        (update report' ::queued-tuples assoc e queue'))
-      report')))
+(defn- transact-report
+  ([report datom] (transact-report report datom false))
+  ([report datom upsert?]
```

Comment:
Why do you use this polymorphism approach instead of multimethods or protocols?

Reply:
In this case I was just looking for a simple way to avoid code duplication to make it clearer what is happening for upserts opposed to inserts. Protocols would not have been applicable since we are not dispatching on type and multimethods would have lead to code duplication again, which I wanted to avoid in the first place. 

For the `with-datom` function we could think about dispatching on the operations `:insert` or `:upsert`. This would make things clearer I'd say.